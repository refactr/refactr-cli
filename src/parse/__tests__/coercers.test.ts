import { join } from 'path';
import { promises as fs } from 'fs';
import {
  coercePipelineCreateInput,
  coerceRunPipelineVariables
} from '../coercers';

describe('coercers', () => {
  describe('coercePipelineCreateInput', () => {
    const fixturePaths = {
      json: join(__dirname, './fixtures/coercePipelineCreateInput/test.json'),
      yaml: join(__dirname, './fixtures/coercePipelineCreateInput/test.yaml')
    };
    let expected: Record<string, unknown>;
    beforeAll(() => {
      expected = require('./fixtures/coercePipelineCreateInput/extected.json');
    });

    test('throws when array is passed', () => {
      expect(() =>
        coercePipelineCreateInput(['@./arg'])
      ).toThrowErrorMatchingSnapshot();
    });

    test('read JSON input string', async () => {
      expect(
        coercePipelineCreateInput(await fs.readFile(fixturePaths.json, 'utf8'))
      ).toStrictEqual(expected);
    });

    test('read YAML input string', async () => {
      expect(
        coercePipelineCreateInput(await fs.readFile(fixturePaths.yaml, 'utf8'))
      ).toStrictEqual(expected);
    });

    test('read JSON file', () => {
      expect(coercePipelineCreateInput(`@${fixturePaths.json}`)).toStrictEqual(
        expected
      );
    });

    test('read YAML file', () => {
      expect(coercePipelineCreateInput(`@${fixturePaths.yaml}`)).toStrictEqual(
        expected
      );
    });
  });

  describe('coerceRunPipelineVariables', () => {
    test('throws on invalid format', () => {
      expect(() =>
        coerceRunPipelineVariables('foo')
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        coerceRunPipelineVariables('foo:')
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        coerceRunPipelineVariables('')
      ).toThrowErrorMatchingSnapshot();
    });

    test('throws on invalid value format', () => {
      expect(() =>
        coerceRunPipelineVariables('foo:bar:')
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        coerceRunPipelineVariables('foo::')
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        coerceRunPipelineVariables('foo:bar:baz')
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        coerceRunPipelineVariables(['foo:bar:baz', 'foo:bar', 'xyz:::'])
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        coerceRunPipelineVariables('foo:bar')
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        coerceRunPipelineVariables('foo:{baz}')
      ).toThrowErrorMatchingSnapshot();
    });

    test('accepts valid JSON value', () => {
      expect(coerceRunPipelineVariables('foo:"bar"')).toStrictEqual({
        foo: 'bar'
      });
      expect(coerceRunPipelineVariables('foo:["bar"]')).toStrictEqual({
        foo: ['bar']
      });
      expect(
        coerceRunPipelineVariables([
          'foo:{"bar":"bar"}',
          'bar:123',
          'baz:"xyz"',
          'xyz:["foo","bar","baz"]'
        ])
      ).toStrictEqual({
        foo: { bar: 'bar' },
        bar: 123,
        baz: 'xyz',
        xyz: ['foo', 'bar', 'baz']
      });
    });
  });
});
