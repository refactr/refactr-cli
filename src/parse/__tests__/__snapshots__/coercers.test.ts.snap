// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`coercers coercePipelineCreateInput throws when array is passed 1`] = `"It is not possible to provide multiple data as input!"`;

exports[`coercers coerceRunPipelineVariables rejects any JSON except object 1`] = `"Invalid variable received \`123\`, expected variable to be specified as \`key:value\` where value must be valid JSON data or valid JSON object."`;

exports[`coercers coerceRunPipelineVariables rejects any JSON except object 2`] = `"Invalid variable received \`false\`, expected variable to be specified as \`key:value\` where value must be valid JSON data or valid JSON object."`;

exports[`coercers coerceRunPipelineVariables rejects any JSON except object 3`] = `"Invalid variable received \`true\`, expected variable to be specified as \`key:value\` where value must be valid JSON data or valid JSON object."`;

exports[`coercers coerceRunPipelineVariables rejects any JSON except object 4`] = `"Invalid variable received \`[]\`, expected variable to be specified as \`key:value\` where value must be valid JSON data or valid JSON object."`;

exports[`coercers coerceRunPipelineVariables throws on invalid format 1`] = `"Invalid variable received \`foo\`, expected variable to be specified as \`key:value\` where value must be valid JSON data or valid JSON object."`;

exports[`coercers coerceRunPipelineVariables throws on invalid format 2`] = `"Invalid variable received \`foo:\`, expected variable to be specified as \`key:value\` where value must be valid JSON data or valid JSON object."`;

exports[`coercers coerceRunPipelineVariables throws on invalid format 3`] = `"Invalid variable received \`\`, expected variable to be specified as \`key:value\` where value must be valid JSON data or valid JSON object."`;

exports[`coercers coerceRunPipelineVariables throws on invalid value format 1`] = `"Invalid variable value received in \`foo:bar:\` pair, expected value to be valid JSON."`;

exports[`coercers coerceRunPipelineVariables throws on invalid value format 2`] = `"Invalid variable value received in \`foo::\` pair, expected value to be valid JSON."`;

exports[`coercers coerceRunPipelineVariables throws on invalid value format 3`] = `"Invalid variable value received in \`foo:bar:baz\` pair, expected value to be valid JSON."`;

exports[`coercers coerceRunPipelineVariables throws on invalid value format 4`] = `"Invalid variable value received in \`foo:bar:baz\` pair, expected value to be valid JSON."`;

exports[`coercers coerceRunPipelineVariables throws on invalid value format 5`] = `"Invalid variable value received in \`foo:bar\` pair, expected value to be valid JSON."`;

exports[`coercers coerceRunPipelineVariables throws on invalid value format 6`] = `"Invalid variable value received in \`foo:{baz}\` pair, expected value to be valid JSON."`;
